contracts/eosio.bios/src/eosio.bios.cpp

```diff
@@ -54,4 +56,122 @@ void bios::reqactivated( const eosio::checksum256& feature_digest ) {
    check( is_feature_activated( feature_digest ), "protocol feature is not activated" );
 }
 
+
+void bios::setresparams(double ram_price, uint64_t total_ram_available) {
+    require_auth("gov.tmy"_n); // check authorization is gov.tmy
+
+    // check ram_price is within bounds, not negative or too high
+    eosio::check(ram_price >= 0, "RAM price must be non-negative");
+
+    // check total_ram_available is non-negative
+    eosio::check(total_ram_available >= 0, "Total RAM available must be non-negative");
+
+    resource_config_table resource_config_singleton(get_self(), get_self().value);
+    resource_config config;
+
+    if (resource_config_singleton.exists()) {
+        // Singleton exists, get the existing config and modify the two values
+        config = resource_config_singleton.get();
+        config.ram_price = ram_price;
+        config.total_ram_available = total_ram_available;
+    } else {
+        // Singleton does not exist, set the two values and also set other values to 0
+        config = resource_config{ram_price, total_ram_available, 0, 0, 0};
+    }
+
+    // Save the modified or new config back to the singleton
+    resource_config_singleton.set(config, get_self());
 }
+
+void bios::buyram(eosio::name dao_owner, eosio::name app, eosio::asset quant) {
+    require_auth(app); // Check that the app has the necessary authorization
+
+   // Access the account table from id.tmy.hpp
+   idtmy::id::account_type_table account_type("id.tmy"_n, "id.tmy"_n.value);
+   // Check the account type of the app
+   auto itr = account_type.find(app.value);
+   eosio::check(itr != account_type.end(), "Could not find account");
+   eosio::check(itr->account_type == idtmy::enum_account_type::App, "Only apps can buy and sell RAM");
+
+   // Check that the RAM is being purchased with the correct token
+   eosio::check(quant.symbol == bios::system_resource_currency, "must buy ram with core token");
+
+   // Check that the amount of tokens being used for the purchase is positive
+   eosio::check(quant.amount > 0, "Amount must be positive");
+
+    // Check the decimal place of the token amount
+    eosio::check(quant.symbol.precision() <= 4, "Token amount must have up to 4 decimal places");
```

Comment:
`quant.symbol.precision() >= 4`
shouldn't it be this? why have you got it as a maximum, not a minimum?